### PR TITLE
MDEV-18851: multiple sized large page support (linux)

### PR DESCRIPTION
**Pull request by grooverdan**
For all the reasons in MDEV-18851, large page support is horrible. This fixes all of these by utilizing the multiple hugetlbs size available size 2.6.32. This uses the mmap/shmget interfaces that have flags for specifying the hugetlbs size that where available in Linux kernel 3.8. The earliest kernel version in the distros supported by MariaDB is 3.12 (SLES12).

As a cleanup, in the later parts of this pull request, the default my_large_page_size and system readonly variable has been removed (as they are no longer used) and other bits of copying large_page enablement where removed. 

Extending the mysys my_large_malloc to use mmap rather than shmget was possible, however it required a rework of the my_large_free interface and needed unravelling of maria storage engine and key_cache to keep track of the size allocated (needed for unmap - as rounded to multiple of pagesize). As such this has been left for a future PR. ([WIP seems to be working](https://github.com/grooverdan/mariadb-server/compare/10.4-MDEV-18851-multiple-sized-large-page-support...10.4-MDEV-18851-maria-my_large_malloc_to_mmap))

I submit this under the MCA.

### CODE REVIEW

File: storage/innobase/os/os0proc.cc [hunk 1]
```diff
@@ -64,44 +61,43 @@ void*
...
 		goto skip;
 	}
 
-	/* Align block size to os_large_page_size */
-	ut_ad(ut_is_2pow(os_large_page_size));
-	size = ut_2pow_round(*n + (os_large_page_size - 1),
-			     os_large_page_size);
+	/* MAP_HUGE_SHIFT added linux-3.8. Take largest HUGEPAGE size */
+	while ((large_page_size = my_next_large_page_size(*n, &i))) {
+		mapflag = MAP_PRIVATE | MAP_ANONYMOUS | MAP_HUGETLB | my_bit_log2(large_page_size) << MAP_HUGE_SHIFT;
```
Comment (dr-m) [inline storage/innobase/os/os0proc.cc]: `my_bit_log2()` takes `ulong`, not `size_t`.
Do we really need both the logarithm and bit-shifting? Could we just reject not-power-of-2 page sizes (which are extremely unlikely in practice) when parsing the data from the kernel? See `ut_is_2pow()`.

File: storage/innobase/os/os0proc.cc [hunk 2]
```diff
@@ -186,4 +176,11 @@ os_mem_free_large(
 		os_total_large_mem_allocated -= size;
 	}
 #endif
+#if defined HAVE_LINUX_LARGE_PAGES && defined UNIV_LINUX
+	if (my_use_large_pages) {
+		/* note accounting will be off if we fell though to
+		conventional memory on allocation */
+		os_total_large_mem_allocated -= size;
+	}
+#endif /* HAVE_LINUX_LARGE_PAGES && UNIV_LINUX */
```
Comment (dr-m) [inline storage/innobase/os/os0proc.cc]: Yes, you just moved some code around, but while doing it you might clean it up.
Doesn’t `HAVE_LINUX_LARGE_PAGES` imply `UNIV_LINUX`?

File: mysys/my_largepage.c [hunk 3]
```diff
@@ -25,21 +25,82 @@
...
+            " errno %d\n", errno);
+  }
+  else
+  {
+    while (i < my_large_page_sizes_length &&
+          (r= readdir(dirp)))
+    {
+      if (strncmp("hugepages-", r->d_name, 10) == 0)
+      {
+        sizes[i]= atoll(r->d_name + 10) * 1024L;
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply mysys/my_largepage.c]: ppc64le has 16G huge pages. I'll look closer at type conversions.

File: storage/innobase/os/os0proc.cc (hunk 1 above)
Comment (grooverdan) [reply storage/innobase/os/os0proc.cc]: > Do we really need both the logarithm and bit-shifting?

Yes. log2(2M) is 20 and there's no way to get from 2M to 20 * 2^^26 without both transforms. Added size_t equivalent of `my_bit_log2`

**Comment by grooverdan** [comment]
Updated based on review comments. Thanks @dr-m.

Also removed a lot more CMake variables no longer used.

Second last commit adds more large page usage to maria.

Note the errors on shutdown, unmapping 1 byte - is fixed in #1220.
```
2019-03-13 11:46:25 0 [ERROR] InnoDB: munmap(0x7fe400000000, 1) failed; errno 22
2019-03-13 11:46:25 0 [ERROR] InnoDB: munmap(0x7fe380000000, 1) failed; errno 22
```

Useful during testing for seeing MMUPageSize/KernelPageSize of differing sizes being used:
```
cd /proc/$(pidof mysqld); egrep -A 20 '(SYSV|huge)' smaps

more /sys/devices/system/node/node*/hugepages/hugepages-*/*
```

Setup instructions
```
# setting 4 1G pages to be available
echo 4 > /sys/devices/system/node/node0/hugepages/hugepages-1048576kB/nr_hugepages
# extend number of default 2M pages
echo 1000 > /sys/devices/system/node/node0/hugepages/hugepages-2048kB/nr_hugepages
```

Suggestions on how to write a mtr test case welcome (none exist for large pages). Read `/sys/devices/system/node/node*/hugepages/hugepages-*kB/free_hugepages` in mysql-test/mysql-test-run.pl and set environment variables like `BIG_TEST`?

File: storage/innobase/os/os0proc.cc [hunk 4]
```diff
@@ -64,50 +66,49 @@ void*
...
-				" errno " << errno;
 			ptr = NULL;
+			if (errno == ENOMEM) {
+				/* no memory at this size, try next size */
+				continue;
+			}
+			ib::warn() << "Failed to mmap HugeTLB memory segment " << adjusted_size
+				<< " bytes, on large page size " << large_page_size
+				<< " bytes. errno " << errno;
+			ptr = NULL;
```
Comment (grooverdan) [inline storage/innobase/os/os0proc.cc]: nit: second setting of ptr=NULL

File: include/my_bit.h [hunk 5]
```diff
@@ -37,6 +37,12 @@ static inline uint my_bit_log2(ulong value)
   return bit;
 }
 
+static inline uint my_bit_size_t_log2(size_t value)
+{
+  uint bit;
+  for (bit=0 ; value > 1 ; value>>=1, bit++) ;
+  return bit;
```
Comment (grooverdan) [inline include/my_bit.h]: would probably be better as type template like https://github.com/MariaDB/server/commit/e7f426d2c9be3ebc7c373e80873996ce8857ab48#diff-28377f2a728b889ab5a14d8e5f31de8eR169 ?

**Comment by grooverdan** [comment]
re-based after significant conflicts from MDEV-19212

**Comment by dr-m** [comment]
@grooverdan Thanks for the update!

I saw an earlier comment from you, suggesting to use type-safe function templates in `my_bit.h`. Actually, I believe that MySQL did that in 5.7 or 8.0. That is where I got the idea for my "prevent the likes of MDEV-19212 from reoccurring" commit e7f426d2c9be3ebc7c373e80873996ce8857ab48.

But, unlike MySQL, some of the MariaDB server is still compiled as C, not C++. @svoj could we maybe define the things twice in `my_bit.h`: as macros when the file is included from C code, and as function templates for C++?

**Comment by grooverdan** [comment]
Came to the same conclusion too. In this update I did change some macros in my_largepage.c to static inline of size_t to preserve the same level of safety.

**Comment by grooverdan** [comment]
So @svoj , see anything wrong here?

**Comment by svoj** [comment]
Re shm to mmap migration: from what I can see the only problem with shm is limits, right? Or are there any other issues? On my xenial `ipcs -lm` says:
```
------ Shared Memory Limits --------
max number of segments = 4096
max seg size (kbytes) = 18014398509465599
max total shared memory (kbytes) = 18014398442373116
min seg size (bytes) = 1
```

Which I read as no limit. Did you actually hit "Failed to attach shared memory segment, errno 22" error?

**Comment by svoj** [comment]
Also reading about transparent huge pages, do we actually benefit anything from supporting regular huge pages?

**Comment by grooverdan** [comment]
On x86_64 4096 segments at a default 2M hugepagesize is only 8G.

Yes shm -> mmap is mostly limits but also shm also required a hugetlb_shm_group group. As a bonus it removes the fidlyness with shm handles which made os_mem_free_large a lot easier as its just an unmap to remove memory, regardless if it was huge or fell back to conventional. 

I don't remember hitting "Failed to attach shared memory segment, errno 22", though it was a while ago. I think I hit EPERM/ENOMEM.

I suspect the reason for shm was that mmap hugetlbs wasn't added until kernel version 2.6.32 and the mysql shm code seems to pre-date this (and RHEL5 was still in support 2.6.18).

Preparing a performance test now (I did design only on pure fast memory access one after watching  Fernando Laudares Carmagos FOSDEM talk). Things like tokudb complain if transparent HugeTLB is enabled so this just enable HugeTLB for maria/innodb.

**Comment by akopytov** [comment]
@grooverdan is right, in 2004 when I was implementing HugeTLB support, 2.4 was the latest stable kernel. So shm was the only option. I also remember reading Oracle docs that mentioned only shm.

**Comment by svoj** [comment]
Interesting, MongoDB manual suggests that transparent huge pages are evil for database workloads:
https://docs.mongodb.com/manual/tutorial/transparent-huge-pages/

**Comment by grooverdan** [comment]
While waiting for @dr-m's approval, https://github.com/grooverdan/mariadb-server/tree/10.5-MDEV-18851-multiple-sized-large-page-support is a WIP (aria tools break linking on Windows and Windows not tested). mtr test case suggestions welcome.

**Comment by grooverdan** [comment]
Re-based on 10.5 as per proposed plan.

Thanks for the interim Windows review @vaintroub

**Comment by grooverdan** [comment]
@svoj Recoded as per your the plan from June.

**Comment by dr-m** [comment]
> Interesting, MongoDB manual suggests that transparent huge pages are evil for database workloads:
> https://docs.mongodb.com/manual/tutorial/transparent-huge-pages/
> 
> I suggest the following:
> 
>     1. Re-target for 10.5.
> 
>     2. Port Windows large pages support from InnoDB to mysys.
> 
>     3. Remove `os_mem_alloc_large()` and `os_mem_free_large()`, use `my_large_malloc()` instead.
> 
>     4. Do shm -> mmap transition (without multiple page sizes support).
> 
>     5. Then we could discuss the rest.
> 
> 
> @dr-m do you approve the above plan?

Yes, I think that it is reasonable to have only one ‘operating system interface’ layer, in the core code base, not separately in each storage engine. Replacing `os_mem_alloc_large()` with `my_large_malloc()` is a welcome step to that direction. Maybe we should also move `aligned_malloc()` to `my_aligned_malloc()`?

On a possibly related note, in [MDEV-21239](https://jira.mariadb.org/browse/MDEV-21239) it looks like AddressSanitizer may get confused by `mmap()` calls. Perhaps for ASAN-instrumented builds, we should map `my_large_malloc()` to the `my_aligned_malloc()` that I proposed above.

File: mysys/my_largepage.c [hunk 6]
```diff
@@ -79,88 +66,354 @@ void my_large_free(uchar* ptr)
...
+}
+#endif
+
+#if defined(HAVE_MMAP) && !defined(_WIN32)
 
-  ptr = (uchar*) shmat(shmid, NULL, 0);
-  if (ptr == (uchar *) -1)
+/* mmap and Linux-specific large pages deallocator */
+
+my_bool my_large_free_int(void *ptr, size_t size)
```
Comment (kevgs) [inline mysys/my_largepage.c]: For manual poisoned memory ASAN requires that it will be unpoisoned manually too. `munmap()` certainly doesn't unpoison any memory. https://github.com/llvm/llvm-project/blob/master/compiler-rt/include/sanitizer/asan_interface.h#L21

IMO, ideally, the guy who poisoned memory should unpoison it too. But he might forget to do it. In that case unpoisoning memory before `munmap()` will help preventing false positives from ASAN.

@grooverdan what do you think, is it worth to unpoison memory here?

Comment (kevgs) [reply mysys/my_largepage.c]: Here is a relevant MariaDB commit https://github.com/MariaDB/server/commit/f6b58d29163adcb3b3fe1b48c56c9b80666f0b88

Comment (grooverdan) [reply mysys/my_largepage.c]: Getting definitions in order:
<pre>UNIV_MEM_FREE -> MEM_NOACCESS -> ASAN_POISON_MEMORY_REGION
UNIV_MEM_ALLOC -> UNIV_MEM_INVALID -> MEM_UNDEFINED -> ASAN_UNPOISON_MEMORY_REGION
</pre>
That MEM_FREE definition really played with my head.

So yes, the current implementation seems to unpoision on allocation and freeing and this should too. I'm going to add that to the current patch sequence in this PR. Thanks for the references.

**Comment by grooverdan** [comment]
rebased and inserted ASAN commit.

Compiled WITH_ASAN;

tests:
<pre>
# setting 4 1G pages to be available
echo 4 > /sys/devices/system/node/node0/hugepages/hugepages-1048576kB/nr_hugepages
# extend number of default 2M pages
echo 1500 > /sys/devices/system/node/node0/hugepages/hugepages-2048kB/nr_hugepages

$ sql/mysqld  --large-pages --innodb-buffer-pool-size=6G --innodb-buffer-pool-instances=2 --innodb-buffer-pool-chunk-size=1G

$ cd /proc/$(pidof mysqld); egrep -A 20 '(SYSV|huge)' smaps
</pre>
show segments of 1G with different MMU sizes.

<pre>
$ mysql-test/mtr --mysqld=--large-pages --mem --parallel=10 --max-test-fail=10 --force 

Failing test(s): main.events_bugs plugins.server_audit sys_vars.large_pages_basic unit.conc_dyncol innodb_gis.rtree_recovery

</pre>

main.events_bugs - unrealted test failure
plugins.server_audit  - passed on re-run
sys_vars.large_pages_basic - changing default meant test failed
unit.conc_dyncol - ASAN failure
innodb_gis.rtree_recovery - didn't repeat on re-run

No attributable errors.

File: mysys/my_largepage.c (hunk 6 above)
Comment (kevgs) [reply mysys/my_largepage.c]: IMO, those `MEM_` AND `UNIV_` macros exists for code obfuscation only..

File: mysys/my_largepage.c [hunk 7]
```diff
@@ -15,53 +15,43 @@
...
   DBUG_ENTER("my_large_malloc");
   
-  if (my_use_large_pages && my_large_page_size)
+  if ((ptr = my_large_malloc_int(size, my_flags)) != NULL)
   {
-    if ((ptr = my_large_malloc_int(size, my_flags)) != NULL)
-        DBUG_RETURN(ptr);
-    if (my_flags & MY_WME)
-      fprintf(stderr, "Warning: Using conventional memory pool\n");
+    MEM_UNDEFINED(ptr, *size);
```
Comment (kevgs) [inline mysys/my_largepage.c]: By fact, on every platform `my_large_malloc_int()` returns zero-filled buffer. I suggest to make it a part of function interface (mention in comment) and do not poison memory here at all.

Comment (kevgs) [reply mysys/my_largepage.c]: Sorry, it's actually unposoning happens here, OK, it doesn't matter because there is no sense in unpoisoning a zero-initialized buffer.

File: mysys/my_largepage.c [hunk 8]
```diff
@@ -79,88 +69,353 @@ void my_large_free(uchar* ptr)
     my_large_malloc_int(), i.e. my_malloc_lock() was used so we should free it
     with my_free_lock()
   */
-  if (!my_use_large_pages || !my_large_page_size || !my_large_free_int(ptr))
+  if (!my_large_free_int(ptr, size))
     my_free_lock(ptr);
+  else
+    MEM_UNDEFINED(ptr, size);
```
Comment (kevgs) [inline mysys/my_largepage.c]: For every byte of memory in a program ASAN remembers whether it's poisoned or not. It has some huge array or something similar. On allocating functions like `malloc()` it unpoisons memory, on freeing function like `free()` it poisons memory. But for functions like `munmap()` it does nothing. If memory was poisoned it will remains so. Then OS may reuse that memory for some TLS or stack variable and for ASAN it will be automatically poisoned. And we'll have hard to debug false positives like in https://jira.mariadb.org/browse/MDEV-21239

In short, someone should unpoison potentially poisoned memory right before or right after (it doesn't matter when) `munmap()` or `VirtualFree()`.

Comment (grooverdan) [reply mysys/my_largepage.c]: Ok. Thanks. Reused your words in the comments.

File: mysys/my_largepage.c [hunk 9]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
     my_large_malloc_int(), i.e. my_malloc_lock() was used so we should free it
     with my_free_lock()
   */
-  if (!my_use_large_pages || !my_large_page_size || !my_large_free_int(ptr))
+  if (!my_large_free_int(ptr, size))
     my_free_lock(ptr);
```
Comment (kevgs) [inline mysys/my_largepage.c]: BTW, what memory we're unlocking here? After successful unmapping it doesn't exist in application. I think it's at least redundant.

Comment (kevgs) [reply mysys/my_largepage.c]: Sorry, I was confused by `my_large_free_int()` return value. We do unlocking only on `munmap()` failure. At least it's correct.

**Comment by grooverdan** [comment]
re-based 6 month ago per guidance from previous review. Updated per ASAN. Can I please get an updated review/merge?

File: include/my_bit.h [hunk 10]
```diff
@@ -38,6 +38,14 @@ static inline uint my_bit_log2(ulong value)
 }
 
 
+static inline uint my_bit_size_t_log2(size_t value)
```
Comment (vaintroub) [inline include/my_bit.h]: you could just modify the function above my_bit_log2() , so it accept size_t . it is always larger or equal type to ulong,
i.e truncations from ulong to size_t do not happen.

File: mysys/my_largepage.c [hunk 11]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+/* Windows-specific large pages deallocator */
 
-my_bool my_large_free_int(uchar *ptr)
+my_bool my_large_free_int(void *ptr, size_t size)
 {
   DBUG_ENTER("my_large_free_int");
-  DBUG_RETURN(shmdt(ptr) == 0);
+  /* When RELEASE memory, the size parameter must be 0.
+     Do not use MEM_RELEASE with MEM_DECOMMIT. */
+  if (!VirtualFree(ptr, 0, MEM_RELEASE))
```
Comment (vaintroub) [inline mysys/my_largepage.c]: passing NULL as first parameter to VirtualFree() is handled as an error, which you might want to catch here, because in other APIs this is usually ok.
I have seen Innodb printing both "VirtualAlloc failed with ERROR_INSUFFICIENT_RESOURCES" followed by "VirtualFree failed with ERROR_INVALID_PARAMETER" on allocation errors.  a little ugly.

**Comment by svoj** [comment]
@grooverdan, sorry about these delays. I need to check with the team (on Tuesday, 18) about destiny of this change. I'll do my best to get it in by 10.5 feature freeze, in other words in coming weeks.

File: include/my_bit.h (hunk 10 above)
Comment (grooverdan) [reply include/my_bit.h]: I couldn't see this in the C11 standard, but happy to accept that is true and put in a compile assertion just to be sure. Thanks for the review.

**Comment by grooverdan** [comment]
Rebased on 10.5 to remove conflicts and made the few minor changes based on @vaintroub 's review.

**Comment by svoj** [comment]
@grooverdan I appreciate it. The XA review is taking more time than it was originally expected. This review is still next in the line.

**Comment by grooverdan** [comment]
Rebased again based on a few more conflicts.

Less sure about [static_assert]([https://github.com/MariaDB/server/pull/1221/commits/b669ff385579c44ec09b06cb69c39cc7d2e3b6c1#diff-ae2918cbb3ca4dc3a1c8d0d358e7f03aR76]) as its the first use in the code base. A compile_assert won't work here because of the CONSTEXPR. Seems c++11 is required in 10.5 which is why I assumed its safe

I looked at the re-instrumentation of memory access in MDEV-16431. For large pages this means we'll need over-allocate a request amount which results in dramatic waste like MDEV-18726, or a significant rework of the structures of every allocation. If there's problems here they could be traced with uprobles.

File: mysys/my_largepage.c [hunk 12]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+/* Descending sort */
+
+static int size_t_cmp(const void *a, const void *b)
 {
-  MYSQL_FILE *f;
-  uint size = 0;
-  char buf[256];
-  DBUG_ENTER("my_get_large_page_size_int");
+    const size_t *ia = (const size_t *)a; // casting pointer types
+    const size_t *ib = (const size_t *)b;
```
Comment (svoj) [inline mysys/my_largepage.c]:
```
const size_t *ib= (const size_t *) b;
```
please fix indentation

File: mysys/my_largepage.c [hunk 13]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+    if (*ib > *ia)
+    {
+         return 1;
```
Comment (svoj) [inline mysys/my_largepage.c]: Please fix indentation.

File: mysys/my_largepage.c [hunk 14]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+
+For first use, have *start=0. There is no need to increment *start.
+
+@param   sz size to be searched for.
+@param   start ptr to int representing offset in my_large_page_sizes to start from.
+*start is updated during search and can be used to search again if 0 isn't returned.
+
+@returns the next size found. *start will be incremented to the next potential size.
+@retval  a large page size that is valid on this system or 0 if no large page size possible.
+*/
```
Comment (svoj) [inline mysys/my_largepage.c]: Please fix indentation with the comment: double space.

File: mysys/my_largepage.c [hunk 15]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+      /* other errors are more serious */
+      DBUG_RETURN(NULL);
+    }
+    else /* success */
+    {
+      if (large_page_size)
+      {
+         /* we do need to record the adjustment so that munmap gets called with
+            the right size. This is only the case for HUGETLB pages.
+            Align block size to large_page_size */
```
Comment (svoj) [inline mysys/my_largepage.c]:
```
/*
  we do need to record the adjustment so that munmap gets called with
  the right size. This is only the case for HUGETLB pages.
  Align block size to large_page_size
*/
```

File: mysys/my_largepage.c [hunk 16]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+}
+
+#endif /* defined(HAVE_LINUX_LARGE_PAGES) || defined(HAVE_MMAP_ALIGNED) */
+
+#if defined(HAVE_GETPAGESIZES) && !defined(HAVE_LINUX_LARGE_PAGES)
+void my_get_large_page_sizes(size_t sizes[my_large_page_sizes_length])
+{
+  int nelem;
+
+  nelem = getpagesizes(NULL, 0);
```
Comment (svoj) [inline mysys/my_largepage.c]:
```
nelem= getpagesizes(NULL, 0);
```

File: mysys/my_largepage.c [hunk 17]
```diff
@@ -79,88 +70,361 @@ void my_large_free(uchar* ptr)
...
+
+/* mmap and Linux-specific large pages deallocator */
+
+my_bool my_large_free_int(void *ptr, size_t size)
+{
+  DBUG_ENTER("my_large_free_int");
+
+# if defined(UNIV_SOLARIS)
+  if (munmap(static_cast<caddr_t>(ptr), size))
+# else
```
Comment (svoj) [inline mysys/my_largepage.c]: IIUC it was needed for some ancient Solaris'es. Now they should happily accept `void*`.

File: sql/mysqld.cc [hunk 18]
```diff
@@ -6625,7 +6623,7 @@ struct my_option my_long_options[]=
    "mysql.gtid_slave_pos",
    &gtid_pos_auto_engines, 0, 0, GET_STR, REQUIRED_ARG,
    0, 0, 0, 0, 0, 0 },
-#ifdef HAVE_LARGE_PAGE_OPTION
+#ifdef HAVE_SOLARIS_LARGE_PAGES
```
Comment (svoj) [inline sql/mysqld.cc]: Unsure about this change. We normally don't remove options (or tighten support) without deprecating them first.

File: sql/sys_vars.cc [hunk 19]
```diff
@@ -1350,16 +1350,10 @@ static Sys_var_mybool Sys_large_files_support(
        READ_ONLY GLOBAL_VAR(opt_large_files),
        CMD_LINE_HELP_ONLY, DEFAULT(sizeof(my_off_t) > 4));
 
-static Sys_var_uint Sys_large_page_size(
-       "large_page_size",
-       "If large page support is enabled, this shows the size of memory pages",
-       READ_ONLY GLOBAL_VAR(opt_large_page_size), NO_CMD_LINE,
-       VALID_RANGE(0, UINT_MAX), DEFAULT(0), BLOCK_SIZE(1));
```
Comment (svoj) [inline sql/sys_vars.cc]: We can't remove it without marking it deprecated first. See how `DEPRECATED()` is used in the sys_vars.cc. Alternatively you can remove it but add to the `removed_opts` of mysqld.cc.

File: storage/innobase/buf/buf0buf.cc [hunk 20]
```diff
@@ -1453,7 +1453,9 @@ static buf_chunk_t* buf_chunk_init(buf_chunk_t* chunk, ulint mem_size)
 	chunk->blocks = (buf_block_t*) chunk->mem;
 
 	/* Align a pointer to the first frame.  Note that when
-	opt_large_page_size is smaller than srv_page_size,
+	large_page_size is smaller than srv_page_size,
+        (with max srv_page_size at 64k don't think any hardware
+        makes this true),
```
Comment (svoj) [inline storage/innobase/buf/buf0buf.cc]: It is InnoDB: please re-indent this comment with tabs.

File: include/my_bit.h [hunk 21]
```diff
@@ -70,6 +70,15 @@ static inline CONSTEXPR uint my_bit_log2_uint64(ulonglong value)
          my_bit_log2_uint32((uint32) (value >> 32)) + 32 :
          my_bit_log2_uint32((uint32) value);
 }
+static inline CONSTEXPR uint my_bit_log2(size_t value)
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply include/my_bit.h]: nope. my fault - lazy non-renaming during rebase past intermediate commit a1e330de5a37

File: include/my_sys.h [hunk 22]
```diff
@@ -176,15 +176,37 @@ extern void *my_memdup(PSI_memory_key key, const void *from,size_t length,myf My
...
+#define my_large_malloc(A,B) my_malloc_lock(*(A),(B))
+#define my_large_free(A,B) my_free_lock((A))
+#endif /* HAVE_LARGE_PAGE_OPTION */
+
+#ifdef _WIN32
+extern BOOL my_obtain_privilege(LPCSTR lpPrivilege);
+#else
+#define my_obtain_privilege(A) (1)
+#define SE_LOCK_MEMORY_NAME "SeLockMemoryPrivilege"
+#endif
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply include/my_sys.h]: I doubted no other OS would every need an my_init_large_pages so I didn't abstract it.

File: include/my_sys.h [hunk 23]
```diff
@@ -176,15 +176,37 @@ extern void *my_memdup(PSI_memory_key key, const void *from,size_t length,myf My
...
 extern char *my_strndup(PSI_memory_key key, const char *from, size_t length, myf MyFlags);
 
-#ifdef HAVE_LINUX_LARGE_PAGES
-extern uint my_get_large_page_size(void);
-extern uchar * my_large_malloc(size_t size, myf my_flags);
-extern void my_large_free(uchar *ptr);
+#if defined(HAVE_LINUX_LARGE_PAGES) || defined(HAVE_GETPAGESIZES)
+#define my_large_page_sizes_length 8
+extern size_t my_large_page_sizes[my_large_page_sizes_length];
+extern void my_get_large_page_sizes(size_t sizes[]);
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply include/my_sys.h]: true, was keeping consistent with the extern function definitions either side of this block.

File: mysys/mf_keycache.c [hunk 24]
```diff
@@ -545,10 +546,17 @@ int init_simple_key_cache(SIMPLE_KEY_CACHE_CB *keycache,
...
+         is the same as the large page size. */
+      if (my_use_large_pages)
+      {
+        keycache->allocated_mem_size= use_mem;
+      }
+      else
+      {
+        keycache->allocated_mem_size= blocks * keycache->key_cache_block_size;
+      }
+      if ((keycache->block_mem=my_large_malloc(&keycache->allocated_mem_size, MYF(0))))
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply mysys/mf_keycache.c]: yes the whole commit is independent and I'll push it to a separate PR.

File: include/my_sys.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (svoj) [reply include/my_sys.h]: Feels like it is needed everywhere already, e.g. to replace `my_get_large_page_size()` call from mysqld.cc

[the remark this answers is not in the document]
Comment (grooverdan) [reply include/my_sys.h]: so it sounds like you want all the mysqld.cc initialisation to be one in a my_init_large_pages function.Ok. I'll get that done tomorrow.

[the remark this answers is not in the document]
Comment (svoj) [reply include/my_sys.h]: It is up to you to decide. If it is not going to be trivial, we can clean it up later.

File: cmake/os/Linux.cmake [hunk 25]
```diff
@@ -42,10 +42,7 @@ ENDIF()
...
 # Linux specific HUGETLB /large page support
-CHECK_SYMBOL_EXISTS(SHM_HUGETLB sys/shm.h  HAVE_DECL_SHM_HUGETLB)
-IF(HAVE_DECL_SHM_HUGETLB)
-  SET(HAVE_LINUX_LARGE_PAGES 1)
-  SET(HUGETLB_USE_PROC_MEMINFO 1)
+CHECK_SYMBOL_EXISTS(MAP_HUGE_SHIFT linux/mman.h HAVE_LINUX_LARGE_PAGES)
+IF(HAVE_LINUX_LARGE_PAGES)
   SET(HAVE_LARGE_PAGE_OPTION 1)
 ENDIF()
-
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply cmake/os/Linux.cmake]: given the end of lifetime extended support dates of https://access.redhat.com/support/policy/updates/errata/#Life_Cycle_Dates I'd rather leave it a little longer.

File: configure.cmake [hunk 26]
```diff
@@ -416,6 +414,10 @@ CHECK_FUNCTION_EXISTS (vsnprintf HAVE_VSNPRINTF)
 CHECK_FUNCTION_EXISTS (memalign HAVE_MEMALIGN)
 CHECK_FUNCTION_EXISTS (nl_langinfo HAVE_NL_LANGINFO)
 
+IF(HAVE_MMAP)
+  SET(HAVE_LARGE_PAGE_OPTION 1)
+ENDIF()
```
[the remark this answers is not in the document]
Comment (grooverdan) [reply configure.cmake]: removed in cmake/os/Linux.cmake.

File: cmake/os/Linux.cmake (hunk 25 above)
[the remark this answers is not in the document]
Comment (svoj) [reply cmake/os/Linux.cmake]: Do you think they use these ancient kernels? Anyway, I doubt 10.5 will even compile on RHEL 5. Probably not even on RHEL 6.

File: include/my_sys.h (hunk 22 above)
[the remark this answers is not in the document]
Comment (grooverdan) [reply include/my_sys.h]: ok done in second last commit.

File: cmake/os/Linux.cmake (hunk 25 above)
[the remark this answers is not in the document]
Comment (grooverdan) [reply cmake/os/Linux.cmake]: good point corrected

**Comment by grooverdan** [comment]
corrected depreciation notices. Testing Linux/FreeBSD. FreeBSD (with other fixes per PRs). FreeBSD ran with `mtr --mysqld=--large-pages`. OpenIndiana managed to compile mysqld.cc/my_large.c with help from #1392 however too many other compile issues exist.

**Comment by svoj** [comment]
Thanks! I'm not sure what has happened when I merged it, but it got squashed. Even though I used "Rebase and merge".
